src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -676,6 +676,234 @@ size_t runtime::gpu::CUDAEmitter::build_elementwise_n_to_1(const GPURuntimeConte
     return primitive_index;
 }
 
+size_t runtime::gpu::CUDAEmitter::build_replace_slice(const GPURuntimeContext* ctx,
+                                                      const std::array<std::string, 3>& dtypes,
+                                                      const Shape& tensor_shape,
+                                                      const Shape& source_shape,
+                                                      const Coordinate& lower_bounds,
+                                                      const Coordinate& upper_bounds,
+                                                      const Strides& slice_strides)
+{
+    // assumes NC{d1,...,dn} format
+    std::string kernel_name = "repslice";
+    std::stringstream ss;
+    ss << kernel_name << "_s" << join(tensor_shape, "_") << "_ssrc" << join(source_shape, "_")
+       << "_sll" << join(lower_bounds, "_") << "_slu" << join(upper_bounds, "_") << "_slst"
+       << join(slice_strides, "_");
+    auto hash = ss.str();
+
+    // check if the requested kernel is already an inserted primitive
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    constexpr const int nthreads_per_block = 32;
+
+    // if the kernel has not been compiled, build it
+    auto compiled_kernel = ctx->compiled_kernel_pool->get(kernel_name);
+    if (compiled_kernel == nullptr)
+    {
+        codegen::CodeWriter writer;
+        writer << include_helpers();
+
+        writer << "extern \"C\" __global__ void cuda_" << kernel_name << "(" << dtypes[0]
+               << "* in, " << dtypes[1] << "* source, " << dtypes[2] << "* out, "
+               << "float alpha, float beta, "
+               << "int* dim_strides, "
+               << "int* dim_magic, "
+               << "int* dim_shift, "
+               << "int* lower_bounds, "
+               << "int* upper_bounds, "
+               << "int* slice_str, "
+               << "int* slice_magic, "
+               << "int* slice_shift, "
+               << "int* dim_source, "
+               << "int* src_strides, "
+               << "int rank,"
+               << "int nthreads"
+               << ")\n";
+        writer.block_begin();
+        {
+            writer << "extern __shared__ int dimensions[];\n";
+            writer << "const int tid = blockDim.x*blockIdx.x + threadIdx.x;\n";
+            writer << "if (tid < nthreads)\n";
+            writer.block_begin();
+            {
+                writer << "int dim_product = tid;\n";
+                writer << "int data_idx = 0;\n";
+                writer << "for (int i = threadIdx.x; i < (rank - 1) * " << nthreads_per_block
+                       << "; i += " << nthreads_per_block << ")\n";
+                writer.block_begin();
+                {
+                    writer << "dimensions[i] = div64(dim_product, dim_magic[data_idx], "
```

Comment:
nit:
is it possible to wrap this to a function with more readable name?

Reply:
Absolutely, I'll try to find something better.